// crownstone-cloud/dist/CrownstoneWebhooks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const toolchain_1 = require("./tools/toolchain");
const Logger_1 = require("./Logger");
const log = Logger_1.Logger(__filename);
class CrownstoneWebhooks {
    constructor(customEndpoint) {
        this.log = log;
        this.toolchain = new toolchain_1.Toolchain();
        this.rest = this.toolchain.getWebhookRequestor(customEndpoint);
    }
    setApiKey(apiKey) {
        this.toolchain.tokenStore.webhooks.api_key = apiKey;
    }
    setAdminKey(adminKey) {
        this.toolchain.tokenStore.webhooks.admin_key = adminKey;
    }
    async isListenerActiveByToken(token) {
        return await this.rest.isListenerActive(token);
    }
    async isListenerActiveByUserId(userId) {
        return await this.rest.isListenerActive('', userId);
    }
    async createListener(userId, token, eventTypes, url) {
        return await this.rest.createListener(userId, token, eventTypes, url);
    }
    async getListeners() {
        return await this.rest.getListeners();
    }
    async removeListenerByToken(token) {
        return await this.rest.deleteListenerByToken(token);
    }
    async removeListenerByUserId(userId) {
        return await this.rest.deleteListenerByUserId(userId);
    }
}
exports.CrownstoneWebhooks = CrownstoneWebhooks;
