// crownstone-cloud/dist/webhooks/sections/listeners.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const webhookApiBase_1 = require("./webhookApiBase");
exports.listeners = {
    isListenerActive: function (token) {
        let endpoint = 'listeners/active';
        return webhookApiBase_1.webhookApiBase._setupRequest('GET', endpoint, {
            data: {
                token: token,
            },
        }, 'query');
    },
    deleteListenersByToken: function (token) {
        let endpoint = 'listeners/token';
        return webhookApiBase_1.webhookApiBase._setupRequest('DELETE', endpoint, {
            data: {
                token: token,
            },
        }, 'query');
    },
    deleteListener: function (listenerId) {
        let endpoint = 'listeners/' + listenerId;
        return webhookApiBase_1.webhookApiBase._setupRequest('DELETE', endpoint);
    },
    createListener: function (listenerData) {
        let endpoint = 'listeners/';
        return webhookApiBase_1.webhookApiBase._setupRequest('POST', endpoint, { data: listenerData }, "body");
    },
    getListeners: function () {
        let endpoint = 'listeners/';
        return webhookApiBase_1.webhookApiBase._setupRequest('GET', endpoint);
    },
};
