// crownstone-cloud/dist/rest/sections/devices.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cloudApiBase_1 = require("./cloudApiBase");
exports.devices = {
    getDevices: function () {
        return cloudApiBase_1.cloudApiBase._setupRequest('GET', '/users/{id}/devices', { data: { filter: { "include": "installations" } } });
    },
    createDevice: function (data) {
        return cloudApiBase_1.cloudApiBase._setupRequest('POST', '/users/{id}/devices', { data: data }, 'body');
    },
    updateDevice: function (deviceId, data) {
        return cloudApiBase_1.cloudApiBase._setupRequest('PUT', '/Devices/' + deviceId, { data: data }, 'body');
    },
    sendTestNotification: function () {
        return cloudApiBase_1.cloudApiBase._setupRequest('POST', '/Devices/{id}/testNotification/');
    },
    deleteDevice: function (deviceId) {
        return cloudApiBase_1.cloudApiBase._setupRequest('DELETE', '/users/{id}/devices/' + deviceId);
    },
    deleteAllDevices: function () {
        return cloudApiBase_1.cloudApiBase._setupRequest('DELETE', '/users/{id}/deleteAllDevices');
    },
    getTrackingNumberInSphere: function (cloudSphereId) {
        return cloudApiBase_1.cloudApiBase._setupRequest('GET', '/Devices/{id}/trackingNumber/', { data: { sphereId: cloudSphereId } }, 'query');
    },
    inSphere: function (cloudSphereId) {
        return cloudApiBase_1.cloudApiBase._setupRequest('POST', '/Devices/{id}/inSphere/', { data: { sphereId: cloudSphereId } }, 'query');
    },
    inLocation: function (cloudSphereId, cloudLocationId) {
        return cloudApiBase_1.cloudApiBase._setupRequest('POST', '/Devices/{id}/inLocation/', { data: { sphereId: cloudSphereId, locationId: cloudLocationId } }, 'query');
    },
    exitLocation: function (cloudSphereId, cloudLocationId) {
        return cloudApiBase_1.cloudApiBase._setupRequest('POST', '/Devices/{id}/exitLocation/', { data: { sphereId: cloudSphereId, locationId: cloudLocationId } }, 'query');
    },
    exitSphere: function (cloudSphereId) {
        return cloudApiBase_1.cloudApiBase._setupRequest('POST', '/Devices/{id}/exitSphere/', { data: { sphereId: cloudSphereId } }, 'query');
    },
};
