// crownstone-cloud/dist/rest/sections/installations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cloudApiBase_1 = require("./cloudApiBase");
exports.installations = {
    getInstallations: function (options = {}) {
        return cloudApiBase_1.cloudApiBase._setupRequest('GET', '/Devices/{id}/installations', options);
    },
    createInstallation: function (appName, data) {
        return cloudApiBase_1.cloudApiBase._setupRequest('POST', '/Devices/{id}/installations?appName=' + appName, { data: data }, 'body');
    },
    updateInstallation: function (installationId, data) {
        return cloudApiBase_1.cloudApiBase._setupRequest('PUT', '/AppInstallations/' + installationId, { data: data }, 'body');
    },
    getInstallation: function (installationId) {
        return cloudApiBase_1.cloudApiBase._setupRequest('GET', '/AppInstallations/' + installationId);
    },
    deleteInstallation: function (installationId) {
        return cloudApiBase_1.cloudApiBase._setupRequest('DELETE', '/Devices/{id}/devices/' + installationId);
    }
};
