// crownstone-cloud/dist/rest/sections/locations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cloudApiBase_1 = require("./cloudApiBase");
exports.locations = {
    getLocations: function () {
        return cloudApiBase_1.cloudApiBase._setupRequest('GET', '/Spheres/{id}/ownedLocations', { data: { filter: { "include": ["sphereOverviewPosition", "presentPeople"] } } });
    },
    createLocation: function (data) {
        return cloudApiBase_1.cloudApiBase._setupRequest('POST', '/Spheres/{id}/ownedLocations', { data: data }, 'body');
    },
    updateLocation: function (cloudLocationId, data) {
        return cloudApiBase_1.cloudApiBase._setupRequest('PUT', '/Spheres/{id}/ownedLocations/' + cloudLocationId, { data: data }, 'body');
    },
    updateLocationPosition: function (data) {
        return cloudApiBase_1.cloudApiBase._setupRequest('POST', '/Locations/{id}/sphereOverviewPosition/', { data: data }, 'body');
    },
    deleteLocation: function (cloudLocationId) {
        return cloudApiBase_1.cloudApiBase._setupRequest('DELETE', '/Spheres/{id}/ownedLocations/' + cloudLocationId);
    },
    deleteLocationPicture: function () {
        return cloudApiBase_1.cloudApiBase._setupRequest('DELETE', '/Locations/{id}/image/');
    },
};
