// crownstone-cloud/dist/rest/sections/spheres.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cloudApiBase_1 = require("./cloudApiBase");
const stones_1 = require("./stones");
const locations_1 = require("./locations");
const cloudAPI_1 = require("../cloudAPI");
exports.spheres = {
    updateSphere: function (cloudSphereId, data) {
        return cloudAPI_1.REST._setupRequest('PUT', '/Spheres/' + cloudSphereId, { data: data }, 'body');
    },
    inviteUser: function (email, permission = "") {
        permission = permission.toLowerCase();
        switch (permission) {
            case 'admin':
                return cloudAPI_1.REST._setupRequest('PUT', '/Spheres/{id}/admins', { data: { email: email } });
            case 'member':
                return cloudAPI_1.REST._setupRequest('PUT', '/Spheres/{id}/members', { data: { email: email } });
            case 'guest':
                return cloudAPI_1.REST._setupRequest('PUT', '/Spheres/{id}/guests', { data: { email: email } });
            default:
                return new Promise((resolve, reject) => {
                    reject(new Error('Invalid Permission: "' + permission + '"'));
                });
        }
    },
    getPendingSphereInvites: function () {
        return cloudAPI_1.REST._setupRequest('GET', '/Spheres/{id}/pendingInvites');
    },
    resendInvite: function (email) {
        return cloudAPI_1.REST._setupRequest('GET', '/Spheres/{id}/resendInvite', { data: { email: email } });
    },
    revokeInvite: function (email) {
        return cloudAPI_1.REST._setupRequest('GET', '/Spheres/{id}/removeInvite', { data: { email: email } });
    },
    /**
     *
     * @returns {*}
     */
    getSpheres: function () {
        return cloudAPI_1.REST._setupRequest('GET', '/users/{id}/spheres', { data: { filter: { include: "floatingLocationPosition" } } });
    },
    getUsers: function () {
        return cloudAPI_1.REST._setupRequest('GET', '/Spheres/{id}/users');
    },
    getAdmins: function () {
        return cloudAPI_1.REST._setupRequest('GET', '/Spheres/{id}/admins');
    },
    getMembers: function () {
        return cloudAPI_1.REST._setupRequest('GET', '/Spheres/{id}/members');
    },
    getGuests: function () {
        return cloudAPI_1.REST._setupRequest('GET', '/Spheres/{id}/guests');
    },
    getToons: function () {
        return cloudAPI_1.REST._setupRequest('GET', '/Spheres/{id}/Toons');
    },
    getPresentPeople: function (ignoreDeviceId) {
        return cloudAPI_1.REST._setupRequest('GET', '/Spheres/{id}/PresentPeople', {
            data: { ignoreDeviceId: ignoreDeviceId },
        }, 'query');
    },
    /**
     * @param data
     */
    createSphere: function (data) {
        return cloudAPI_1.REST._setupRequest('POST', 'users/{id}/spheres', { data: data }, 'body');
    },
    changeSphereName: function (sphereName) {
        return cloudAPI_1.REST._setupRequest('PUT', '/Spheres/{id}', { data: { name: sphereName } }, 'body');
    },
    changeUserAccess: function (email, accessLevel) {
        return cloudAPI_1.REST._setupRequest('PUT', '/Spheres/{id}/role', { data: { email: email, role: accessLevel } }, 'query');
    },
    updateFloatingLocationPosition: function (data) {
        return cloudAPI_1.REST._setupRequest('POST', '/Spheres/{id}/floatingLocationPosition/', { data: data }, 'body');
    },
    deleteUserFromSphere: function (userId) {
        // userId is the same in the rest as it is locally
        return cloudAPI_1.REST._setupRequest('DELETE', '/Spheres/{id}/users/rel/' + userId);
    },
    deleteSphere: function () {
        let sphereId = cloudApiBase_1.TokenStore.sphereId;
        let promises = [];
        let stoneData = [];
        let locationData = [];
        promises.push(stones_1.stones.getStonesInSphere()
            .then((stones) => {
            stoneData = stones;
        }).catch((err) => { }));
        // for every sphere, we get the locations
        promises.push(locations_1.locations.getLocations()
            .then((locations) => {
            locationData = locations;
        }).catch((err) => { }));
        return Promise.all(promises)
            .then(() => {
            let deletePromises = [];
            stoneData.forEach((stone) => {
                deletePromises.push(cloudAPI_1.REST.forSphere(sphereId).deleteStone(stone.id));
            });
            locationData.forEach((location) => {
                deletePromises.push(cloudAPI_1.REST.forSphere(sphereId).deleteLocation(location.id));
            });
            return Promise.all(deletePromises);
        })
            .then(() => {
            return this._deleteSphere(sphereId);
        });
    },
    _deleteSphere: function (cloudSphereId) {
        if (cloudSphereId) {
            return cloudAPI_1.REST._setupRequest('DELETE', 'Spheres/' + cloudSphereId);
        }
    },
    leaveSphere: function (cloudSphereId) {
        if (cloudSphereId) {
            return cloudAPI_1.REST._setupRequest('DELETE', 'users/{id}/spheres/rel/' + cloudSphereId);
        }
    },
    acceptInvitation: function () {
        return cloudAPI_1.REST._setupRequest('POST', '/Spheres/{id}/inviteAccept/', {}, 'body');
    },
    declineInvitation: function () {
        return cloudAPI_1.REST._setupRequest('POST', '/Spheres/{id}/inviteDecline/', {}, 'body');
    },
    getSphereAuthorizationTokens: function (sphereId) {
        if (sphereId) {
            cloudApiBase_1.TokenStore.sphereId = sphereId;
        }
        return cloudAPI_1.REST._setupRequest('GET', 'Spheres/{id}/tokenData');
    }
};
