// homey/fn/dispatch.js
/*!
 * homey v0.1.0-beta.13
 * https://github.com/demiazz/homey
 *
 * Copyright 2017-present Alexey Plutalov <demiazz.py@gmail.com>
 * Released under the MIT license
 */

"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _utils = require("../utils");

function createWithConstructor(eventType, detail, bubbles, cancelable) {
  return new CustomEvent(eventType, {
    bubbles: bubbles,
    cancelable: cancelable,
    detail: detail
  });
}

function createWithInit(eventType, detail, bubbles, cancelable) {
  var event = document.createEvent("CustomEvent");

  event.initCustomEvent(eventType, bubbles, cancelable, detail);

  return event;
}

var create =
  typeof window.CustomEvent === "function"
    ? createWithConstructor
    : createWithInit;

function dispatch(target, eventType) {
  var detail =
    arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  var options =
    arguments.length > 3 && arguments[3] !== undefined
      ? arguments[3]
      : { bubbles: true, cancelable: true };

  var bubbles = (0, _utils.getProperty)(options, "bubbles", true);
  var cancelable = (0, _utils.getProperty)(options, "cancelable", true);
  var event = create(eventType, detail, bubbles, cancelable);

  return target.dispatchEvent(event);
}

exports.default = dispatch;
