// crownstone-cloud/dist/dataContainers/crownstone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Crownstone {
    constructor(cloudRequestor, stoneId) {
        this.stoneId = null;
        this.rest = cloudRequestor;
        this.stoneId = stoneId;
    }
    async data() {
        return await this.rest.getCrownstone(this.stoneId);
    }
    async currentSwitchState() {
        let data = await this.rest.getCurrentSwitchState(this.stoneId);
        let switchState = data.switchState || 0;
        if (switchState > 0 && switchState <= 1) {
            return switchState * 100;
        }
        return data.switchState;
    }
    async setCurrentSwitchState(switchState) {
        await this.rest.setCurrentSwitchState(this.stoneId, switchState);
    }
    async setSwitch(value) {
        await this._switch({ type: "PERCENTAGE", percentage: value });
    }
    async turnOn() {
        await this._switch({ type: "TURN_ON" });
    }
    async turnOff() {
        await this._switch({ type: "TURN_OFF" });
    }
    async _switch(switchData) {
        let value = switchData.percentage;
        // normalize value
        if (value) {
            if (value > 0 && value <= 1) {
                value *= 100;
            }
            value = Math.max(0, Math.min(100, value));
        }
        await this.rest.switchCrownstone(this.stoneId, { type: switchData.type, percentage: value });
    }
}
exports.Crownstone = Crownstone;
