// crownstone-cloud/dist/dataContainers/sphere.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Sphere {
    constructor(cloudRequestor, id = null) {
        this.sphereId = null;
        this.rest = cloudRequestor;
        this.sphereId = id;
    }
    async users() {
        return await this.rest.getUsers(this.sphereId);
    }
    async authorizationTokens() {
        return await this.rest.getSphereAuthorizationTokens(this.sphereId);
    }
    async crownstones() {
        return await this.rest.getCrownstonesInSphere(this.sphereId);
    }
    async locations() {
        return await this.rest.getLocationsInSphere(this.sphereId);
    }
    async keys() {
        let keys = await this.rest.getKeys();
        for (let i = 0; i < keys.length; i++) {
            if (keys[i].sphereId === this.sphereId) {
                return keys[i];
            }
        }
        return null;
    }
    async data() {
        return this.rest.getSphere(this.sphereId);
    }
    async presentPeople(ignoreDeviceId) {
        return this.rest.getPresentPeople(this.sphereId, ignoreDeviceId);
    }
}
exports.Sphere = Sphere;
