// crownstone-cloud/dist/tools/requestorBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class RequestorBase {
    constructor(tokenStore, cache) {
        this.tokenStore = tokenStore;
    }
    setEndpoint(endpoint) {
        if (endpoint[endpoint.length - 1] !== '/') {
            this.endpoint = endpoint + '/';
        }
        else {
            this.endpoint = endpoint;
        }
    }
    addSecurity(options) {
        if (this.tokenStore.accessToken) {
            if (this.tokenStore.accessToken.length > 32) {
                return { ...options, headers: { ...options.headers, Authorization: this.tokenStore.accessToken } };
            }
            else {
                return { ...options, searchParams: { ...options.searchParams, access_token: this.tokenStore.accessToken } };
            }
        }
        return options;
    }
    get hookSecurityAdmin() {
        if (this.tokenStore.webhooks.admin_key) {
            return { headers: { admin_key: this.tokenStore.webhooks.admin_key } };
        }
        return {};
    }
    get hookSecurityApi() {
        if (this.tokenStore.webhooks.api_key) {
            return { headers: { api_key: this.tokenStore.webhooks.api_key } };
        }
        return {};
    }
}
exports.RequestorBase = RequestorBase;
