// crownstone-cloud/dist/tools/toolchain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tokens_1 = require("./tokens");
const requestors_1 = require("./requestors");
const cache_1 = require("./cache");
class Toolchain {
    constructor() {
        this.tokenStore = new tokens_1.TokenStore();
        this.cache = new cache_1.CacheStorage();
    }
    loadUserData(email, passwordSha1) {
        this.tokenStore.cloudUser.email = email;
        this.tokenStore.cloudUser.passwordSha1 = passwordSha1;
    }
    loadAccessToken(accessToken, userId) {
        this.tokenStore.accessToken = accessToken;
        if (userId) {
            this.tokenStore.cloudUser.userId = userId;
        }
    }
    loadHubData(hubId, hubToken) {
        this.tokenStore.cloudHub.hubId = hubId;
        this.tokenStore.cloudHub.hubToken = hubToken;
    }
    loadHubSphereId(sphereId) {
        this.tokenStore.cloudHub.sphereId = sphereId;
    }
    getCloudRequestor(customEndpoint) {
        return new requestors_1.CloudRequestor(this.tokenStore, this.cache, customEndpoint).interface();
    }
    getWebhookRequestor(customEndpoint) {
        return new requestors_1.WebhookRequestor(this.tokenStore, this.cache, customEndpoint).interface();
    }
}
exports.Toolchain = Toolchain;
