// crownstone-core/dist/packets/ResultPacket.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Wrapper for all relevant data of the object
 *
 */
const DataStepper_1 = require("../util/DataStepper");
const CrownstoneTypes_1 = require("../protocol/CrownstoneTypes");
class ResultPacket {
    constructor(data) {
        this.valid = false;
        this.load(data);
    }
    load(data) {
        let minSize = 7;
        if (data.length >= minSize) {
            this.valid = true;
            let stepper = new DataStepper_1.DataStepper(data);
            this.protocol = stepper.getUInt8();
            this.commandType = stepper.getUInt16();
            this.resultCode = stepper.getUInt16();
            if (CrownstoneTypes_1.ControlTypeInv[this.commandType] === undefined || CrownstoneTypes_1.ResultValueInv[this.resultCode] === undefined) {
                this.valid = false;
                return;
            }
            this.size = stepper.getUInt16();
            let totalSize = minSize + this.size;
            if (data.length >= totalSize) {
                this.payload = stepper.getBuffer(this.size);
            }
            else {
                this.valid = false;
            }
        }
        else {
            this.valid = false;
        }
    }
}
exports.ResultPacket = ResultPacket;
