// crownstone-core/dist/containers/CrownstoneSettings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Util_1 = require("../util/Util");
class CrownstoneSettings {
    constructor() {
        this.encryptionEnabled = true;
        this.adminKey = null;
        this.memberKey = null;
        this.basicKey = null;
        this.serviceDataKey = null;
        this.localizationKey = null;
        this.meshNetworkKey = null;
        this.meshAppKey = null;
        this.setupKey = null;
        this.initializedKeys = false;
        // determineUserLevel() {
        //   if (this.adminKey.length == 16) {
        //     this.userLevel = UserLevel.admin
        //   }
        //   else if (this.memberKey.length == 16) {
        //     this.userLevel = UserLevel.member
        //   }
        //   else if (this.basicKey.length == 16) {
        //     this.userLevel = UserLevel.basic;
        //   }
        //   else {
        //     this.userLevel = UserLevel.unknown;
        //     this.initializedKeys = false
        //   }
        // }
    }
    loadKeys(encryptionEnabled, adminKey = null, memberKey = null, basicKey = null, serviceDataKey = null, localizationKey = null, meshNetworkKey = null, meshAppKey = null, referenceId) {
        this.encryptionEnabled = encryptionEnabled;
        this.adminKey = this._prepKey(adminKey);
        this.memberKey = this._prepKey(memberKey);
        this.basicKey = this._prepKey(basicKey);
        this.serviceDataKey = this._prepKey(serviceDataKey);
        this.localizationKey = this._prepKey(localizationKey);
        this.meshNetworkKey = this._prepKey(meshNetworkKey);
        this.meshAppKey = this._prepKey(meshAppKey);
        this.initializedKeys = true;
    }
    _prepKey(key) {
        return Util_1.Util.prepareKey(key);
    }
}
exports.CrownstoneSettings = CrownstoneSettings;
